syntax = "proto3";
package pain.manusya;

import "core/common.proto";

option cc_generic_services = true;

message CreateChunkRequest {};
message CreateChunkResponse {
  UUID uuid = 1;
};

message AppendChunkRequest {
  UUID uuid = 1;
  uint64 offset = 2;
  uint32 length = 3;
  uint32 digest = 4;
};

message AppendChunkResponse {};

message SealChunkRequest {};
message SealChunkResponse {};

message RemoveChunkRequest {};
message RemoveChunkResponse {};

message ListChunkRequest {
  UUID start = 1;
  uint32 limit = 2;
};

message ListChunkResponse {
  repeated UUID uuids = 1;
};

message ReadChunkRequest {
  UUID uuid = 1;
  uint64 offset = 2;
  uint32 length = 3;
};

message ReadChunkResponse {
  uint64 offset = 1;
  uint32 length = 2;
  uint32 digest = 3;
};

service ManusyaService {
  rpc create_chunk(CreateChunkRequest) returns (CreateChunkResponse);
  rpc append_chunk(AppendChunkRequest) returns (AppendChunkResponse);
  rpc seal_chunk(SealChunkRequest) returns (SealChunkResponse);
  rpc remove_chunk(RemoveChunkRequest) returns (RemoveChunkResponse);
  rpc list_chunk(ListChunkRequest) returns (ListChunkResponse);
  rpc read_chunk(ReadChunkRequest) returns (ReadChunkResponse);
};
